### PR TITLE
[ml/train] `Trainer` [2/4]: `DataParallel` implementation

### PR DESCRIPTION
**Pull request by amogkam**
Initial implementation for training on Ray AIR part 2.

Adds `DataParallelTrainer` to Ray AIR, which is built on top of the current Ray Train.

Depends on
https://github.com/ray-project/ray/pull/22921

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/checkpoint.py [hunk 1]
```diff
@@ -1,117 +1,224 @@
...
-
-import ray
-
-
-class Checkpoint(abc.ABC):
-    """Checkpoint interface"""
-
-    def __init__(self, metadata: Optional[Dict] = None):
-        self.metadata = metadata or {}
+import shutil
```
Comment (ericl) [inline python/ray/ml/checkpoint.py:1]: Hm assuming this checkpoint thing will go away?

File: python/ray/ml/config.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+# Right now, RunConfig is just an arbitrary dict that specifies tune.run
+# kwargs.
+# TODO: After Tuner is implemented, make this into an actual dataclass
+RunConfig = Dict[str, Any]
+
+ScalingConfig = Dict[str, Any]
+
+
+@dataclass
+class _DataParallelScalingConfig:
```
Comment (ericl) [inline python/ray/ml/config.py:16]: Move this to the data parallel class definition?

File: python/ray/train/impl/data_parallel_trainer.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+if TYPE_CHECKING:
+    import ray.data.Dataset
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelFunctionTrainer(Trainer):
+    """
```
Comment (ericl) [inline python/ray/train/impl/data_parallel_trainer.py:30]: Remove newline after """

File: python/ray/train/impl/data_parallel_trainer.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    A Trainer for data parallel training.
+
+    This Trainer runs the provided function on multiple Ray Actors. The
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    Args:
+        train_func (Callable): The training function to execute.
```
Comment (ericl) [inline python/ray/train/impl/data_parallel_trainer.py:38]:
```suggestion
        train_func: The training function to execute.
```

Here and elsewhere.

File: python/ray/train/impl/data_parallel_trainer.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            backend to setup on the workers enable distributed
+            communication, for example torch or horovod. If no Backend
+            should be set up, then set this to None.
+        scaling_config (Optional[ScalingConfig]): Configuration
+            for how to scale data parallel training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or :ref:`DatasetPipeline
+            <dataset-pipeline-api>`, or a Callbable that returns a Dataset, to
```
Comment (ericl) [inline python/ray/train/impl/data_parallel_trainer.py:52]:
```suggestion
            <dataset-pipeline-api>`, or a Callable that returns a Dataset to
```

File: python/ray/train/impl/data_parallel_trainer.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            it will be fit on this dataset.
+        additional_datasets (Optional[GenDataset]): Any additional Ray
+            Datasets (such as validation or test datasets) to use for
+            training. If a ``preprocessor`` is provided, it will only
+            transform these datasets.
+        preprocessor (Optional[Preprocessor]): A preprocessor to preprocess
+            the provided datasets.
+        resume_from_checkpoint (Optional[Checkpoint]): A checkpoint to
+            resume training from.
+
```
Comment (ericl) [inline python/ray/train/impl/data_parallel_trainer.py:63]:
```suggestion
```

File: python/ray/train/session.py [hunk 7]
```diff
@@ -336,6 +337,19 @@ def train_func():
     return shard
 
 
+def get_train_dataset_shard() -> Optional[RayDataset]:
```
Comment (ericl) [inline python/ray/train/session.py:340]: RayDataset is not a ray.data.Dataset right?

File: python/ray/ml/checkpoint.py (hunk 1 above)
Comment (amogkam) [reply python/ray/ml/checkpoint.py:1]: Yep, this is just a subset copied from @krfricke's PR.

File: python/ray/ml/config.py (hunk 2 above)
Comment (krfricke) [reply python/ray/ml/config.py:16]: Agreed, that makes this cleaner (see also comments on PR 1/4). My proposal here would be something like:

```

class Trainer(abc.ABC):

    @dataclass
    class ScalingConfig:
        """Documentation for trainer-specific scaling config"""
        pass

    def __init__(self, scaling_config: Union[ScalingConfig, Dict[str, Any]]):
        self._scaling_config = scaling_config

    def _get_scaling_config_class(self, config: Dict) -> ScalingConfig:
        scaling_config = self.ScalingConfig(**self._scaling_config)
        deep_update(scaling_config.__dict__, config.get("scaling_config", {}))
        return scaling_config



class DataParallelTrainer(Trainer):

    @dataclass
    class ScalingConfig:
        num_workers: int = 1
        num_cpus_per_worker: int = 1
        num_gpus_per_worker: int = 0
```

**Pull request closed without merging**